packages/core-data/src/test/reducer.js

```diff
@@ -99,6 +107,206 @@ describe( 'entities', () => {
 	} );
 } );
 
+describe( 'undo', () => {
+	let lastEdits;
+	let undoState;
+	let expectedUndoState;
+	const createEditActionPart = ( edits ) => ( {
+		kind: 'someKind',
+		name: 'someName',
+		recordId: 'someRecordId',
+		edits,
+	} );
+	const createNextEditAction = ( edits, transientEdits = {} ) => {
+		let action = {
+			...createEditActionPart( edits ),
+			transientEdits,
+		};
+		action = {
+			type: 'EDIT_ENTITY_RECORD',
+			...action,
+			meta: {
+				undo: { ...action, edits: lastEdits },
+			},
+		};
+		lastEdits = { ...lastEdits, ...edits };
+		return action;
+	};
+	const createNextUndoState = ( ...args ) => {
+		let action = {};
+		if ( args[ 0 ] === 'isUndo' || args[ 0 ] === 'isRedo' ) {
+			// We need to "apply" the undo level here and build
+			// the action to move the offset.
+			lastEdits =
+				undoState[
+					undoState.length + undoState.offset - ( args[ 0 ] === 'isUndo' ? 2 : 0 )
+				].edits;
+			action = {
+				type: 'EDIT_ENTITY_RECORD',
+				meta: {
+					[ args[ 0 ] ]: true,
+				},
+			};
+		} else if ( args[ 0 ] === 'isCreate' ) {
+			action = { type: 'CREATE_UNDO_LEVEL' };
+		} else if ( args.length ) {
+			action = createNextEditAction( ...args );
+		}
+		return deepFreeze( undo( undoState, action ) );
+	};
+	beforeEach( () => {
+		lastEdits = {};
+		undoState = undefined;
+		expectedUndoState = [];
+		expectedUndoState.offset = 0;
```

Comment:
This is interesting to see in a state, as normally the state should be serialisable. Could you comment on this in the reducer? Why can't it be nested?

Reply:
It could be nested. I just followed from the previous iteration and I think I've seen other places where this is done.

I guess we should change it in another PR if we need to serialize it.